lib/charms/prometheus_k8s/v1/prometheus.py

```diff
@@ -317,99 +314,71 @@ def __init__(self, charm, name, consumes, multi=False):
-                on which the scrapped endpoint exposes its Prometheus
-                metrics.
-            rel_id: an optional integer providing the relation ID for
-                the related Prometheus monitoring service
-                provider. This is only necessary if the
-                `PrometheusConsumer` has been instantiated in
-                `multi` mode.
-        """
-        if rel_id is None:
-            rel_id = self.relation_id
-
-        targets = self._stored.targets.get(rel_id, [])
-        target = address + ":" + str(port)
-        if target not in targets:
-            return
+        event.relation.data[self._charm.app]["prometheus_scrape_metadata"] = json.dumps(
+            self._scrape_metadata
+        )
 
-        targets.remove(target)
-        self._update_targets(targets, rel_id)
+    def _set_unit_ip(self, event):
+        """Set unit host address
 
-    @property
-    def endpoints(self, rel_id=None):
-        """Returns a list of Prometheus scrape target endpoints.
-        Args:
-            rel_id: an optional integer providing the relation ID for
-                the related Prometheus monitoring service
-                provider. This is only necessary if the
-                `PrometheusConsumer` has been instantiated in
-                `multi` mode.
+        Each time a consumer charm container is restarted it updates its own
+        host address in the unit relation data for the Prometheus provider.
         """
+        for relation in self._charm.model.relations[self._relation_name]:
+            relation.data[self._charm.unit]["prometheus_scrape_host"] = str(
+                self._charm.model.get_binding(relation).network.bind_address
+            )
 
-        if rel_id is None:
-            rel_id = self.relation_id
-        return self._stored.targets.get(rel_id, [])
-
-    def _set_targets(self, event):
-        """Set the Prometheus scrape targets."""
-        rel_id = event.relation.id
-        if not self._stored.targets.get(rel_id, []):
-            return
-
-        logger.debug("Setting scrape targets : %s", self._stored.targets[rel_id])
-        event.relation.data[self._charm.app]["targets"] = json.dumps(
-            list(self._stored.targets[rel_id])
-        )
-
-    def _update_targets(self, targets, rel_id):
-        """Update the Prometheus scrape targets."""
-        self._stored.targets[rel_id] = targets
-        rel = self.framework.model.get_relation(self._relation_name, rel_id)
+    @property
+    def _scrape_metadata(self):
+        """Generate scrape metadata.
 
-        logger.debug("Updating scrape targets to : %s", targets)
-        rel.data[self._charm.app]["targets"] = json.dumps(list(targets))
+        Returns:
+            Scrape configutation metadata for this Prometheus consumer charm.
+        """
+        metadata = {
+            "model": "{}".format(self._charm.model.name),
+            "application": "{}".format(self._charm.model.app.name),
```

Comment:
Just be aware that, with CMR, the model and application name could be the same for multiple related applications, but that should be fine with how it's handled because the relation ID is still guaranteed to be unique.

Reply:
There is actually a commit in this PR that intentionally removes relation ID from the list of scrape target labels. This was done on request of @mmanciop because there was concern here that Alertmanager will not be able to dedupe alerts in the presence of relation ID. I think this is a valid concern. A future PR will add model UUID once https://github.com/canonical/operator/pull/563 gets merged.

Reply:
It's merged now.

Reply:
Thanks. Added commit with model uuid into this PR.